corehq/apps/sms/tests/test_form_session_handler.py

```diff
@@ -0,0 +1,136 @@
+from datetime import datetime
+
+from django.test import TestCase
+from mock import patch, Mock, MagicMock
+
+from corehq.apps.domain.models import Domain
+from corehq.apps.sms.handlers.form_session import form_session_handler
+from corehq.apps.sms.models import (
+    PhoneNumber,
+    SMS,
+    INCOMING,
+    MessagingEvent,
+    MessagingSubEvent
+)
+from corehq.apps.sms.tests.util import mock_critical_section_for_smsforms_sessions
+from corehq.apps.smsforms.models import SQLXFormsSession
+from corehq.messaging.smsbackends.test.models import SQLTestSMSBackend
+
+
+@patch('corehq.apps.smsforms.util.critical_section_for_smsforms_sessions',
+       new=mock_critical_section_for_smsforms_sessions)
+class FormSessionTestCase(TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.domain = Domain(name='test')
+        cls.number = PhoneNumber(
+            domain=cls.domain.name,
+            owner_doc_type='CommCareCase',
+            owner_id='fake-owner-id1',
+            phone_number='01112223333',
+            backend_id=None,
+            ivr_backend_id=None,
+            verified=True,
+            is_two_way=True,
+            pending_verification=False,
+            contact_last_modified=datetime.utcnow()
+        )
+        cls.session = SQLXFormsSession.create_session_object(
+            cls.domain.name,
+            Mock(get_id=cls.number.owner_id),
+            cls.number.phone_number,
+            Mock(get_id='app_id'),
+            Mock(xmlns='xmlns'),
+            expire_after=24 * 60,
+        )
+        cls.session.save()
+
+        cls.backend = SQLTestSMSBackend.objects.create(
+            name='BACKEND',
+            domain=cls.domain,
+            is_global=False,
+            hq_api_id=SQLTestSMSBackend.get_api_id()
+        )
+
+    def _create_message_event_and_subevent(self):
+        event = MessagingEvent.objects.create(
+            domain=self.domain.name,
+            date=datetime.utcnow(),
+            source=MessagingEvent.SOURCE_BROADCAST,
+            content_type=MessagingEvent.CONTENT_SMS_SURVEY,
+            status=MessagingEvent.STATUS_COMPLETED
+        )
+        subevent = MessagingSubEvent.objects.create(
+            parent=event,
+            date=datetime.utcnow(),
+            recipient_type=MessagingEvent.RECIPIENT_CASE,
+            content_type=MessagingEvent.CONTENT_SMS_SURVEY,
+            status=MessagingEvent.STATUS_IN_PROGRESS,
+            xforms_session=self.session
+        )
+
+        return event, subevent
```

Comment:
This method mutates database state, but does not cleanup that state at the end of each test that calls it. Might consider adding
```suggestion
        self.addCleanup(event.delete)  # cascades to subevent
        return event, subevent
```